### ISSUE TITLE
Suppressing CS0618 obsolete when referencing from a obsolete member's attribute?

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#51071]
Let's say I have a Entity Framework model like this:

```c#
    [ForeignKey(nameof(User))]
    public int UserID { get; set; }
    public User User { get; set; }
```

Now I would like to change to use another implementation of user so I marked the `User` class, `User` property, and `UserID` property in the above class as `[Obsolete]`. The reference of the `User` class in the `User` property no longer shows the CS0618 warning because the property itself is also obsolete, but the `nameof(User)` still shows the warning, although the `UserID` is also obsolete. I would expect this warning to go off.

Is there any reason it was not designed as so?

_Originally posted by @acaly in https://github.com/dotnet/csharplang/discussions/4391_

**Comment by jaredpar** [comment]
Bit unclear what is being asked here. Can we see the code sample with `[Obsolete]` that the you feel is behaving incorrectly

**Comment by RikkiGibson** [comment]
@acaly does this accurately represent your scenario? [SharpLab](https://sharplab.io/#v2:EYLgtghgzgLgpgJwDQBMQGoA+ABATARgFgAoE7AZgAI9KBBGGBSkOhhAS2AFd4SBvEpSHUq9RgAps+AAyUoASkp9KAXxJrSxANoB5YFAD2AGzjwAumSo0AqlERLVJS9VyUAwksHDd+46bgWxMKUWmII4gB2EGBwBgBm4raI8vKBwRTUACyUALLiigJBwhoqQA===)

**Comment by acaly** [comment]
@RikkiGibson Yes, that's exactly the problem I am having. Thanks for helping me to clarify that.

**Comment by jaredpar** [comment]
The compiler is consistent here, the `[Obsolete]` attribute only removes obsolete warnings for the method implementation, not any attribute usage on the method. Here is a different demo that I think is a bit more direct:

```c#
using System;

[Obsolete]
public class Attr : Attribute { }

public class C {
    [Obsolete]
    [Attr] // Obsolete warning
    public void M() {
        Attr local = null; // no obsolete warning on Attr
    }
}
```

The reasoning here is that the obsolete attribute here impacts the members, and any uses within it. It does not impact any other attributes used on that member. Notice if that you put `[Obsolete]` on `C` all the warnings go away. 

As to why the language chose this over also applying to the attributes I don't have any direct insight here. That decision long predates me being on the team (probably back to C# 1.0). I did verify this behavior goes back to the native compiler though. I suspect it has to do with avoiding attribute circularity issues. 

In any case though, I don't think we'll be changing the behavior at this point so closing as "By Design".

**Comment by acaly** [comment]
@jaredpar 

No, this is not consistent.

```c#
[Obsolete]
public class TestAttribute : Attribute {}

[Obsolete]
[Test] //No warning
public class C1 {
    public void M() {}
}

public class C2 {
    [Obsolete]
    [Test] //Warning
    public void M() {}
}
```

**Comment by acaly** [comment]
There are more examples:

```c#
[Obsolete]
public class TestAttribute : Attribute {}

public class C {
    
    [Obsolete]
    [Test] //No warning
    public class C1 {
        public void M() {}
    }

    [Obsolete]
    [Test] //Warning
    public void M() {}

    [Obsolete]
    [return: Test] //Warning
    public void M(
        [Test] int a //warning
    )
    {
    }

    [Obsolete]
    [return: Test] //Warning
    public void M<T>(
        [Test] int a //No warning
    )
    {
    }
    
    [Obsolete]
    public class C2 {
        [Test] //No warning
        public void M() {}
    }
    
    [Obsolete]
    public string X
    {
        get => nameof(TestAttribute); //No warning
        [Test] //Warning
        set {}
    }
}
```
